### PR TITLE
Make addTx and getCurrentSlot run on the server.

### PR DESCRIPTION
**Pull request by raduom**
# Move block chain state to the node server.

As part of the effort to remove the full chain from the PAB and only use it in debugging mode I decided, first, to remove the `ChainState` from all the front-facing services and move it into the mocked node server (partially accomplished by this PR). 

The `blockReaper` now can run on the server to trim the blockchain (or remove it if we are using a value of `0` for the kept block size).

The old node client has changed in the following ways:
* The node server is started and a `serverHandler` is provided to be used by the `slotCoordinator` and `blockReaper` services.
* The `randomTx` service still makes use of the client interface to the node server to send a new randomly generated transaction. Generating the Tx needs access to the UTXO which is still kept in the node client (to be removed in the next PR).
* The `currentSlot` is tracked by the socket based node client. This is a good idea since the server will send the slot information to the client during synchronisation.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [x] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by raduom** [comment]
Fixes #SCP-1982 and #SCP-1983.

File: plutus-pab/src/Cardano/Wallet/Mock.hs [hunk 1]
```diff
@@ -125,14 +127,21 @@ processWalletEffects trace nodeClientEnv chainIndexEnv mVarState action = do
 runWalletEffects ::
      MonadIO m
     => Trace m WalletMsg -- ^ trace for logging
-    -> ClientEnv -- ^ node client
+    -> Client.ClientHandler -- ^ node client
     -> ClientEnv -- ^ chain index client
     -> Wallets -- ^ current state
     -> Eff (WalletEffects m) a -- ^ wallet effect
+-- <<<<<<< HEAD
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Wallet/Mock.hs:134]: merge artifacts

File: plutus-pab/src/Cardano/Protocol/Socket/Server.hs [hunk 2]
```diff
@@ -74,49 +77,66 @@ data ServerHandler = ServerHandler {
 data ServerCommand =
     -- This command will add a new block by processing
     -- transactions in the memory pool.
-    ProcessBlocks [Block]
+    ProcessBlock
+  | AddTx Tx
+  | TrimTo Int
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Server.hs:82]: Comment

File: plutus-pab/src/Cardano/Protocol/Socket/Client.hs [hunk 3]
```diff
@@ -31,19 +32,26 @@ import           Cardano.Protocol.Socket.Type
...
 -- | Client handler, used to communicate with the client thread.
-data ClientHandler = ClientHandler {
-        chInputQueue :: TQueue Tx,
-        chSocketPath :: FilePath,
-        chThreadId   :: ThreadId
+data ClientHandler = ClientHandler
+    { chInputQueue  :: TQueue Tx
+    , chSocketPath  :: FilePath
+    , chThreadId    :: ThreadId
+    , chCurrentSlot :: MVar Slot
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Client.hs:39]: Shouldn't this be a `TVar`? I saw that `getCurrentSlot` uses `takeMVar`, so it leaves it empty until a new slot has started. (Also I can't see where `chCurrentSlot` is written to)

**Review by j-mueller — COMMENTED** [review]
Some commented out / merge code that needs to be deleted, otherwise LGTM!

File: plutus-pab/src/Cardano/Protocol/Socket/Client.hs [hunk 4]
```diff
@@ -51,7 +51,7 @@ getCurrentSlot ::
      ClientHandler
   -> IO Slot
 getCurrentSlot ClientHandler { chCurrentSlot } =
-    takeMVar chCurrentSlot
+    readMVar chCurrentSlot
```
[the remark this answers is not in the document]
Comment (raduom) [reply plutus-pab/src/Cardano/Protocol/Socket/Client.hs:54]: The `MVar` is initialised here: https://github.com/raduom/plutus/blob/d93fac47babf2b3d6ba43604ebed733c398b8512/plutus-pab/src/Cardano/Protocol/Socket/Client.hs#L64

.. and updated here: https://github.com/raduom/plutus/blob/d93fac47babf2b3d6ba43604ebed733c398b8512/plutus-pab/src/Cardano/Protocol/Socket/Client.hs#L132-L135

We get the nice property from this code that the callbacks are called serialised and in sequence, which is a nice property to have. We may need to reconsider if things are too slow at some point, but I would like to leave it as-is for now.

**Review by j-mueller — APPROVED** [review]

**Comment by raduom** [comment]
I have tested if this change affects the `ChainIndex` service, and as far as I can see everything works now (there was a problem with the client failing to connect, but I made it reconnect on failure).

The second service that is using the client is the wallet, but when I tried to create a wallet on the `master` branch I failed to do so. The REST call just hangs, so I could not test if my changes actually work, because it does not work on master.

I would suggest we merge this one, close the related tickets, add a ticket to investigate what is happening with the Wallet, and make the remaining ticket be blocked by this new one.

**Review by j-mueller — APPROVED** [review]
I agree. Let's merge this and I'll look into the wallet API issue.

**Pull request merged**